### PR TITLE
[DOCS] Neural network Deployment Guide with System Module Mode #1523

### PR DESCRIPTION
**Pull request by srkreddy1238**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

File: docs/deploy/nnvm.md [hunk 1]
```diff
@@ -116,3 +116,22 @@ int main()
     return 0;
 }
 ```
+
+# Deploy as System Module
```
Comment (tqchen) [inline docs/deploy/nnvm.md]: need to be second level header as it is a subsection

**Comment by tqchen** [comment]
cc @qingyuanxingsi does this clarify your question? Please suggest what can be made to further clarify things

**Comment by qingyuanxingsi** [comment]
@tqchen @srkreddy1238 I'm not clear about how to add the generated .o file into my project using a **bazel** like build system, can you give a bazel build file example

I have finished the following:
* Add the tvm runtime library(bazel target)
* Generated the .o file
* cpp_deploy.cc is Ready

**Comment by srkreddy1238** [comment]
Did you try adding the .o file to the bazel ```srcs```

**Comment by qingyuanxingsi** [comment]
@srkreddy1238 tried. 

```
Don't know how to build from a source file with suffix `.o'
```

**Comment by qingyuanxingsi** [comment]
@srkreddy1238 Actually we are using a build system like bazel, but not bazel. I'm trying to make it work!

**Comment by qingyuanxingsi** [comment]
I tried to pack the .o file to .a file using:

```
ar -rcs libexample.a *.o
```

It compiled, but it failed to run, got following error:

```tvm/src/runtime/./graph/graph_runtime.cc:577: Check failed: pf != nullptr no such function in module: fuse_conv2d_broadcast_mul_broadcast_add_relu
terminate called after throwing an instance of 'dmlc::Error'
  what():  [20:22:33] antispam/3rdlib/tvm/src/runtime/./graph/graph_runtime.cc:577: Check failed: pf != nullptr no such function in module: fuse_conv2d_broadcast_mul_broadcast_add_relu
```

Is it ok if I do it like this??

**Comment by srkreddy1238** [comment]
@qingyuanxingsi 

for target=llvm ```lib.get_source()```  and for non targets ```lib.imported_modules[0].get_source()``` can give the llvm source. you may need additional setup to compile on your environment.

I have checked bazel with below changes and the object embeds into the final binary very well.

Add an additional target deploy_resnet18 to pack the nnvm system lib object into library and make deploy_resnet18 dependency for cpp_deploy as shown below. 

Note ```alwayslink=1``` is the trick which enforce embedding this lib completely in final binary even though there is no symbol dependency.

```bash
cc_library(
    name = "deploy_resnet18",
    srcs = ["deploy_resnet18.o"],
    alwayslink=1
)

cc_binary(
    name = "cpp_deploy",
    srcs = ["cpp_deploy.cc"],
    deps = [
        ":tvm_runtime", ":deploy_resnet18"
    ],
)
```

Hope this helps.

**Comment by srkreddy1238** [comment]
@qingyuanxingsi  
about time we are trying options together :)  and you are very close.

I too tried archive of object but that doesn't work alone as the cpp_deploy doesn't refer any symbol directly.
As described above embed the archive rule into the BUILD with ```alwayslink=1``` and try.

**Comment by qingyuanxingsi** [comment]
@srkreddy1238 Have you tried running the app, can you get correct prediction result?
If so, I will give it another try next week. (**The build system we use is a little different from bazel, we may need certain adjustment**) Much thanks for your instructions!!!

**Comment by srkreddy1238** [comment]
I checked upto the symbols of system object in the final binary. Beyond this there shouldn't be any problems.

Pls try and let me know. I will add a section for bazel.

**Comment by srkreddy1238** [comment]
@qingyuanxingsi  

I did a quick check by adding all objects to a bazel build and it work perfectly fine.

```bash
cc_library(
    name = "test_addone_sys",
    srcs = ["test_addone_sys.o"],
    alwayslink=1
)

cc_library(
    name = "tvm_runtime",
    srcs = ["libtvm_runtime_pack.o"],
)

cc_binary(
    name = "bazel_deploy",
    srcs = ["cpp_deploy.o"],
    deps = [
        ":tvm_runtime", ":test_addone_sys"
    ],
    linkopts = [ "-lpthread -ldl" ]
)
```

**Comment by merrymercy** [comment]
@qingyuanxingsi Your error message usually means you don't export the GPU code correctly.
I am not familiar with cpp deployment. But I think "*.o" only works for cpu backend. You should use
`lib.export_library`  (not `lib.save`) to export the library for gpu.

**Comment by tqchen** [comment]
Thanks for the helpful discussion, I think it indeed reveals something that we did not document well, see my comments on https://github.com/dmlc/tvm/issues/1523

**Comment by srkreddy1238** [comment]
@tqchen & @qingyuanxingsi  welcome for review.

File: docs/deploy/nnvm.md [hunk 2]
```diff
@@ -116,3 +116,60 @@ int main()
...
+C++ additionally support deployment as system module.
+This process need few additional options as given below to NNVM build.
+
+- For target llvm append --system-lib as ```target=llvm --system-lib```
+- For a GPU build (or non llvm) the additional option should be given to targat_host as ```target_host=llvm --system-lib```
+
+Module export require additional options for not to compile but save as ```lib.export_library (path, fcompile=False)```
+
+The output of above API is a tar compressed file containing object file ```(lib.o)``` and cpp source file ```(devc.cc)``` which embeds device blob. Thease two files should be compiled along with other files or objects while building c++ application.
+Please refer to [Makefile](https://github.com/dmlc/tvm/blob/54ca1493bd9808a481ca0e7ed32897c12f36796e/apps/howto_deploy/Makefile#L32) for a reference.
```
[the remark this answers is not in the document]
Comment (srkreddy1238) [reply docs/deploy/nnvm.md]: Done.

**Comment by tqchen** [comment]
Thanks @srkreddy1238 @qingyuanxingsi , this is now merged

**Comment by qingyuanxingsi** [comment]
Working now, much thanks. @srkreddy1238 @tqchen

**Pull request merged**